runtime/src/accounts_index.rs

```diff
@@ -881,11 +878,24 @@ impl<T: 'static + Clone> AccountsIndex<T> {
         self.roots_tracker.read().unwrap().roots.contains(&slot)
     }
 
-    pub fn add_root(&self, slot: Slot) {
+    pub fn add_root(&self, slot: Slot, caching_enabled: bool) {
         let mut w_roots_tracker = self.roots_tracker.write().unwrap();
         w_roots_tracker.roots.insert(slot);
-        w_roots_tracker.uncleaned_roots.insert(slot);
-        w_roots_tracker.max_root = std::cmp::max(slot, w_roots_tracker.max_root);
+        // we delay cleaning until flushing!
+        if !caching_enabled {
+            w_roots_tracker.uncleaned_roots.insert(slot);
+        }
+        // `AccountsDb::flush_accounts_cache()` relies on roots being added in order
+        assert!(slot >= w_roots_tracker.max_root);
```

Comment:
nits: should be `slot > w_roots_tracker.max_root`? I mean, no need to be `>=` because the same slot could (and should) never be added twice?

Reply:
hehe this is subtle, but the same slot can actually be added twice, which is why this is a `HashSet`

```
pub fn squash(&self) {
        self.freeze();

        //this bank and all its parents are now on the rooted path
        let mut roots = vec![self.slot()];
        roots.append(&mut self.parents().iter().map(|p| p.slot()).collect());

        let mut squash_accounts_time = Measure::start("squash_accounts_time");
        for slot in roots.iter().rev() {
            // root forks cannot be purged
            self.rc.accounts.add_root(*slot);
        }
```

Specifically this line ` roots.append(&mut self.parents().iter().map(|p| p.slot()).collect());`, multiple banks can have overlapping parents, for instance if you create banks 1-32, then squash bank 2, all banks 2 to 31 will still have bank 2 as a parent.